gma/src/android/native_ad_internal_android.cc

```diff
@@ -0,0 +1,139 @@
+  firebase::MutexLock lock(mutex_);
+  JNIEnv* env = ::firebase::gma::GetJNI();
+  // Since it's currently not possible to destroy the native ad, just
+  // disconnect from it so the listener doesn't initiate callbacks with stale
+  // data.
+  env->CallVoidMethod(
+      helper_, native_ad_helper::GetMethodId(native_ad_helper::kDisconnect));
+  env->DeleteGlobalRef(helper_);
+  helper_ = nullptr;
+}
+
+Future<void> NativeAdInternalAndroid::Initialize(AdParent parent) {
+  firebase::MutexLock lock(mutex_);
+
+  if (initialized_) {
+    const SafeFutureHandle<void> future_handle =
+        future_data_.future_impl.SafeAlloc<void>(kNativeAdFnInitialize);
+    Future<void> future = MakeFuture(&future_data_.future_impl, future_handle);
+    CompleteFuture(kAdErrorCodeAlreadyInitialized,
+                   kAdAlreadyInitializedErrorMessage, future_handle,
+                   &future_data_);
+    return future;
+  }
+
+  initialized_ = true;
+  JNIEnv* env = ::firebase::gma::GetJNI();
+  FIREBASE_ASSERT(env);
+
+  FutureCallbackData<void>* callback_data =
+      CreateVoidFutureCallbackData(kNativeAdFnInitialize, &future_data_);
+  Future<void> future =
+      MakeFuture(&future_data_.future_impl, callback_data->future_handle);
+  env->CallVoidMethod(
+      helper_, native_ad_helper::GetMethodId(native_ad_helper::kInitialize),
+      reinterpret_cast<jlong>(callback_data), parent);
+  return future;
+}
+
+Future<AdResult> NativeAdInternalAndroid::LoadAd(const char* ad_unit_id,
+                                                 const AdRequest& request) {
+  firebase::MutexLock lock(mutex_);
+
+  if (!initialized_) {
+    SafeFutureHandle<AdResult> future_handle =
+        CreateFuture<AdResult>(kNativeAdFnLoadAd, &future_data_);
+    Future<AdResult> future =
+        MakeFuture(&future_data_.future_impl, future_handle);
+    CompleteFuture(kAdErrorCodeUninitialized, kAdUninitializedErrorMessage,
+                   future_handle, &future_data_, AdResult());
+    return future;
+  }
+
+  gma::AdErrorCode error = kAdErrorCodeNone;
+  jobject j_request = GetJavaAdRequestFromCPPAdRequest(request, &error);
+  if (j_request == nullptr) {
+    if (error == kAdErrorCodeNone) {
+      error = kAdErrorCodeInternalError;
+    }
+    return CreateAndCompleteFutureWithResult(
+        kNativeAdFnLoadAd, error, kAdCouldNotParseAdRequestErrorMessage,
+        &future_data_, AdResult());
+  }
+  JNIEnv* env = GetJNI();
+  FIREBASE_ASSERT(env);
+  FutureCallbackData<AdResult>* callback_data =
+      CreateAdResultFutureCallbackData(kNativeAdFnLoadAd, &future_data_);
+  Future<AdResult> future =
+      MakeFuture(&future_data_.future_impl, callback_data->future_handle);
+
+  jstring j_ad_unit_str = env->NewStringUTF(ad_unit_id);
+  ::firebase::gma::GetJNI()->CallVoidMethod(
+      helper_, native_ad_helper::GetMethodId(native_ad_helper::kLoadAd),
+      reinterpret_cast<jlong>(callback_data), j_ad_unit_str, j_request);
```

Comment:
Add `util::CheckAndClearJniExceptions(env);`after call to `::firebase::gma::GetJNI`

Reply:
Thanks! Added it to clear the exceptions.